### PR TITLE
IGVF-736-non-vitual-samples-audit

### PR DESCRIPTION
**Pull request by KAndreeva**
Added an audit to flag non-virtual samples linked to virtual samples.

### CODE REVIEW

File: src/igvfd/audit/sample.py [hunk 1]
```diff
@@ -35,3 +35,53 @@ def audit_sample_sorted_fraction_parent_child_check(value, system):
...
+
+
+def get_virtual_sample_failures(
+    system,
+    sample_id,
+    sample_is_virtual,
+    linked_sample_id
+):
+    if not linked_sample_id:
+        return None
```
Comment (gabdank) [inline src/igvfd/audit/sample.py:69]: this will never happen, you can remove

File: src/igvfd/tests/fixtures/schemas/primary_cell.py [hunk 2]
```diff
@@ -8,19 +8,21 @@ def primary_cell(testapp, other_lab, award, human_donor, sample_term_pluripotent
...
-        'biosample_term': sample_term_pluripotent_stem_cell['@id']
+        'biosample_term': sample_term_pluripotent_stem_cell['@id'],
+        'virtual': True
     }
     return testapp.post_json('/primary_cell', item, status=201).json['@graph'][0]
 
 
 @pytest.fixture
-def pooled_from_primary_cell(testapp, lab, award, source, human_donor, sample_term_pluripotent_stem_cell):
+def pooled_from_primary_cell(testapp, lab, award, source, human_donor, sample_term_pluripotent_stem_cell, primary_cell):
```
Comment (gabdank) [inline src/igvfd/tests/fixtures/schemas/primary_cell.py:18]: why do we add "primary_cell" to the parameters?

File: src/igvfd/tests/fixtures/schemas/primary_cell.py [hunk 3]
```diff
@@ -8,19 +8,21 @@ def primary_cell(testapp, other_lab, award, human_donor, sample_term_pluripotent
         'lab': other_lab['@id'],
         'source': other_lab['@id'],
         'donors': [human_donor['@id']],
-        'biosample_term': sample_term_pluripotent_stem_cell['@id']
+        'biosample_term': sample_term_pluripotent_stem_cell['@id'],
+        'virtual': True
```
Comment (gabdank) [inline src/igvfd/tests/fixtures/schemas/primary_cell.py:12]: why do you make primary_cell virtual?

**Review by gabdank — CHANGES_REQUESTED** [review]

File: src/igvfd/audit/sample.py [hunk 4]
```diff
@@ -35,3 +35,46 @@ def audit_sample_sorted_fraction_parent_child_check(value, system):
...
+                    if audit_failure:
+                        yield audit_failure
+            else:
+                audit_failure = raise_audit_failure_if_not_match_virtual(
+                    system, sample_id, sample_is_virtual, current_property_value, current_property_name)
+                if audit_failure:
+                    yield audit_failure
+
+
+def raise_audit_failure_if_not_match_virtual(system, sample_id, sample_is_virtual, current_property_id, current_property_name):
```
[the remark this answers is not in the document]
Comment (KAndreeva) [reply src/igvfd/audit/sample.py]: Function name changed.

File: src/igvfd/audit/sample.py (hunk 1 above)
Comment (KAndreeva) [reply src/igvfd/audit/sample.py:69]: This change causes the unit test to fail.  The lines on 46 to 48 potentially return None in the list of linked IDs if the property is not present in the sample.

File: src/igvfd/tests/fixtures/schemas/primary_cell.py (hunk 3 above)
Comment (KAndreeva) [reply src/igvfd/tests/fixtures/schemas/primary_cell.py:12]: This is needed to support my unit test named: test_non_virtual_sample_linked_to_virtual_sample_with_single_property

The fixture primary_cell_with_part_of_virtual_true has part_of linked to the fixture primary_cell.  primary_cell_with_part_of_virtual_true is set to virtual False.  primary_cell is set to virtual True.

File: src/igvfd/tests/fixtures/schemas/primary_cell.py (hunk 2 above)
Comment (KAndreeva) [reply src/igvfd/tests/fixtures/schemas/primary_cell.py]: Removed.

File: src/igvfd/tests/fixtures/schemas/primary_cell.py (hunk 3 above)
Comment (gabdank) [reply src/igvfd/tests/fixtures/schemas/primary_cell.py:12]: this fixture is being used by other tests, are you sure they are ok with it becoming virtual?

File: src/igvfd/audit/sample.py (hunk 1 above)
Comment (gabdank) [reply src/igvfd/audit/sample.py:69]: if the linked IDs are not present then links_to_check should be empty list?

File: src/igvfd/tests/fixtures/schemas/primary_cell.py (hunk 3 above)
Comment (gabdank) [reply src/igvfd/tests/fixtures/schemas/primary_cell.py:12]: this fixture is being used by other tests, are you sure they are ok with it becoming virtual?

**Review by gabdank — CHANGES_REQUESTED** [review]

File: src/igvfd/tests/fixtures/schemas/primary_cell.py (hunk 3 above)
Comment (KAndreeva) [reply src/igvfd/tests/fixtures/schemas/primary_cell.py:12]: All unit tests passed with this change, but I can make another fixture if you want?

File: src/igvfd/audit/sample.py (hunk 1 above)
Comment (KAndreeva) [reply src/igvfd/audit/sample.py:69]: But we are populating them with value None if the property is not present. Each entry will be an element of the list.

File: src/igvfd/tests/fixtures/schemas/primary_cell.py (hunk 3 above)
Comment (KAndreeva) [reply src/igvfd/tests/fixtures/schemas/primary_cell.py]: I have removed the virtual property from the primary_cell fixture and patched it in the unit test instead.

File: src/igvfd/audit/sample.py (hunk 1 above)
Comment (KAndreeva) [reply src/igvfd/audit/sample.py:69]: I tried the unit tests with these lines removed, and the test fails.

**Review by gabdank — APPROVED** [review]

**Review by gabdank — APPROVED** [review]

**Pull request merged**